File: randology-dump-master/thesis/src/uk/ac/stir/randology/generator/QCG.java
package uk.ac.stir.randology.generator;

/**
 * An implementation of a Linear Congruential Generator with an arbitrary parameter set.
 * @author zak
 */
public class QCG implements RNG {
	
	private String name;
	
	private long multiplier;
	private long increment;
	private long seed;
	private long modulus;
	
	public QCG(long multiplier, long increment, long modulus, long seed) {
		this("QCG-" + multiplier + "-" + increment + "-" + modulus,multiplier,increment,modulus,seed);
	}
	
	public QCG(String name, long multiplier, long increment, long modulus, long seed) {
		this.multiplier = multiplier;
		this.increment = increment;
		this.modulus = modulus;
		this.seed = seed;
		setName(name);
	}
	
	public String getName() {
		return name;
	}
	
	private void setName(String name) {
		this.name = name;
	}
	
	public long nextLong() {
		seed = (seed*seed*multiplier + seed*(multiplier+1) + increment) % modulus;
		return seed;
	}
	
    public long next64bits() {
    	long high = nextLong();
    	long med  = nextLong();
    	long low  = nextLong();
    	long result = (high << 33) ^ (med << 2) ^ (low << 62 >> 62);
    	return result;
    }
	
	public double nextDouble() {
		return Math.abs( (double)nextLong() / modulus );
	}

}

File: python-implementation/src/randology/pnrg/CSG/QCG.py
import numpy as np

from randology.pnrg import RNG


class QCG(RNG):
    def __init__(self, a, b, increment, modulus, seed):
        super().__init__("QCG{}{}{}-{}".format(a, b, increment, modulus))
        self.a = np.int64(a)
        self.b = np.int64(b)
        self.increment = np.int64(increment)
        self.modulus = np.int64(modulus)
        self.seed = np.int64(seed) % modulus

    def next_int(self) -> np.int64:
        self.seed = (self.seed * self.seed * self.a + self.seed * self.b + self.increment) % self.modulus
        return self.seed

    def next_float(self) -> np.float64:
        return abs(self.next_int() / self.modulus)

    def next_64_bits(self) -> np.int64:
        high = self.next_int()
        med = self.next_int()
        low = self.next_int()
        result = (high << 33) ^ (med << 2) ^ (low << 62 >> 62)
        return result
